ostd/src/arch/x86/trap/gdt.rs

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: MPL-2.0 OR MIT
+
+//! Configure Global Descriptor Table (GDT).
+//!
+//! The most contents of this file come from https://github.com/rcore-os/trapframe-rs.
+
+use alloc::{boxed::Box, vec::Vec};
+use core::mem::size_of;
+
+use x86_64::{
+    instructions::tables::{lgdt, load_tss, sgdt},
+    registers::{
+        model_specific::Star,
+        segmentation::{Segment64, GS},
+    },
+    structures::{
+        gdt::{Descriptor, SegmentSelector},
+        tss::TaskStateSegment,
+        DescriptorTablePointer,
+    },
+    PrivilegeLevel, VirtAddr,
+};
+
+/// Init TSS & GDT.
+pub unsafe fn init(on_bsp: bool) {
+    // Allocate stack for trap from user, set the stack top to TSS,
+    // so that when trap from ring3 to ring0, CPU can switch stack correctly.
+    let tss = if on_bsp {
+        init_local_tss_on_bsp()
+    } else {
+        init_local_tss_on_ap()
+    };
+
+    let (tss0, tss1) = match Descriptor::tss_segment(tss) {
+        Descriptor::SystemSegment(tss0, tss1) => (tss0, tss1),
+        _ => unreachable!(),
+    };
+    // FIXME: the segment limit assumed by x86_64 does not include the I/O port bitmap.
+
+    // Get current GDT.
+    let gdtp = sgdt();
+    let entry_count = (gdtp.limit + 1) as usize / size_of::<u64>();
+    let old_gdt = core::slice::from_raw_parts(gdtp.base.as_ptr::<u64>(), entry_count);
+
+    // Allocate new GDT with 7 more entries.
+    //
+    // NOTICE: for fast syscall:
+    //   STAR[47:32] = K_CS   = K_SS - 8
+    //   STAR[63:48] = U_CS32 = U_SS32 - 8 = U_CS - 16
+    let mut gdt = Vec::from(old_gdt);
+    gdt.extend([tss0, tss1, KCODE64, KDATA64, UCODE32, UDATA32, UCODE64].iter());
+    let gdt = Vec::leak(gdt);
+
+    // Load new GDT and TSS.
+    lgdt(&DescriptorTablePointer {
+        limit: gdt.len() as u16 * 8 - 1,
+        base: VirtAddr::new(gdt.as_ptr() as _),
+    });
+    load_tss(SegmentSelector::new(
+        entry_count as u16,
+        PrivilegeLevel::Ring0,
+    ));
+
+    let sysret = SegmentSelector::new(entry_count as u16 + 4, PrivilegeLevel::Ring3).0;
+    let syscall = SegmentSelector::new(entry_count as u16 + 2, PrivilegeLevel::Ring0).0;
+    Star::write_raw(sysret, syscall);
+
+    USER_SS = sysret + 8;
+    USER_CS = sysret + 16;
+}
+
+// The linker script ensure that cpu_local_tss section is
+// at the beginning of cpu_local, so that gsbase -> LOCAL_TSS.
+#[allow(dead_code)]
+#[link_section = ".cpu_local_tss"]
+static mut LOCAL_TSS: TaskStateSegment = TaskStateSegment::new();
```

Comment:
The semantics of `.cpu_local_tss` is exactly the same as the current `cpu_local` implementation, you can even define it with:

```rust
cpu_local! {
    static LOCAL_TSS: UnsafeCell<TaskStateSegment> = UnsafeCell:new(TaskStateSegment::new());
}
```

You can use the `LOCAL_TSS.as_ptr()` method to get the address of the task state segment on the current CPU.

Also, use an unsafe cell instead of a static mutable.